server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -193,6 +201,23 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         reconcile(currentDesiredBalance, allocation);
     }
 
+    private synchronized void processNodeShutdowns(ClusterState clusterState) {
+        // Clean up processed shutdowns that are removed from the cluster metadata
+        processedNodeShutdowns.removeIf(nodeId -> clusterState.metadata().nodeShutdowns().contains(nodeId) == false);
```

Comment:
If we remove a shutdown marker from a node while the node is still in the cluster, I think that needs a reset too.